### PR TITLE
feat(notification-center): smaller bundle size

### PR DESCRIPTION
**Pull request by LetItRock**
### What change does this PR introduce?

Related: https://github.com/novuhq/novu/issues/2379

The smaller bundle size for the `notification-center` package and dependent `vue, angular` packages.

**The changes:**
- tree shaking for the library and its dependencies
- removed the `date-fns` library and implemented the same logic with the `Intl.RelativeTimeFormat`
- removed the `chroma-js` library
- updated images to webp
- changed the target to ES6

---

**RESULTS IN SHORT (more detailed in screenshots section)**:
1. The UMD Web Component: was ~490.19kB gzip, now ~249.92kB gzip, **~240.27kb gzip less, ~49% smaller bundle**
2. The Example React 17 app using components: was ~355.02kB, now ~201.8kB gzip, **~153.22kB gzip less, ~43% smaller bundle** (ESM)
3. The Example React 17 app using hooks: was ~355.02kB, now ~94.03kB gzip, **~260.99kB gzip less, ~73% smaller bundle** (ESM)
4. The Example Vue app: was ~1414.13kB, now ~715.59kB, **~698.54kB gzip less, ~49% smaller bundle** (ESM)
5. The Example Angular app: was ~1510kB gzip, now ~676.65kB gzip, **~833.35kB gzip less, ~55% smaller bundle** (ESM)

---

**Further improvements that we can do in the future:**
- get rid of the Mantine library
- move the different language objects to the separate tree-shakable package
- use `fetch` instead of Axios because it also requires NodeJS env specific polyfills?

### Other information (Screenshots)

**More detailed:**

---

The UMD Web Component:
- before: ~490.19kB gzip
- after: ~249.92kB gzip
- the difference: **~240.27kb, ~49% smaller bundle**

The Web Component App
Screenshot 2023-06-19 at 17 08 58

Before Web Component:
Screenshot 2023-06-15 at 17 15 26

After Web Component:
Screenshot 2023-06-19 at 14 49 27

---

The ESM module (CJS should be similar):
- before: ~56.07kB gzip
- after: ~44.12kB gzip
- the difference: **~11.95kB, ~21% smaller module size** only because the images were changed to webp

Before:
Screenshot 2023-06-15 at 17 17 47

After:
Screenshot 2023-06-19 at 14 48 51

---

The example React 17 application bundled with Vite
- before: ~355.02kB gzip
- after with component: ~201.8kB gzip
- the difference: **~153.22kB, ~43% smaller bundle**

- after with only hooks used in the code: ~94.03kB gzip
- the difference: **~260.99kB gzip, ~73% smaller bundle**

Before:
Screenshot 2023-06-18 at 17 12 07

After with `notification-center` components:
Screenshot 2023-06-19 at 14 42 53

After with `notification-center` hooks:
Screenshot 2023-06-19 at 14 38 27

---

The Next.js App using the `notification-center` components
- before: ~374kB
- after: ~242kB
- the difference: **~132kb, ~35% smaller bundle**

Screenshot 2023-06-19 at 15 37 50

Before:
Screenshot 2023-06-19 at 15 41 08

After:
Screenshot 2023-06-19 at 15 39 34

---

The Novu Web Application
- before: ~2.11MB gzip
- after: ~1.98MB gzip
- the difference: **~0.13MB, ~6% smaller bundle** :D

Before:
Screenshot 2023-06-18 at 16 29 41

After:
Screenshot 2023-06-19 at 14 47 43

Confirming images are working:
Screenshot 2023-06-19 at 12 58 33

---

The `@novu/notification-center-vue` package
- before ESM:  ~531.18kB gzip
- after: ~364.22kB gzip
- the difference: **~166.96kB, ~31% smaller bundle**

Before:
Screenshot 2023-06-19 at 14 57 16

After:
Screenshot 2023-06-19 at 14 56 28

An example Vue app
- before:  ~1414.13kB gzip
- after: ~715.59kB gzip
- the difference: **~698.54kB, ~49% smaller bundle**

Screenshot 2023-06-19 at 15 19 52

Before:
Screenshot 2023-06-19 at 15 22 10

After: 
Screenshot 2023-06-19 at 15 20 48

---

The `@novu/notification-center-angular` package
- before: ???
- after: ???
- the difference: ???

An example Angular app
- before:  ~1510kB gzip
- after: ~676.65kB gzip
- the difference: **~833.35kB, ~55% smaller bundle**

Screenshot 2023-06-19 at 15 27 53

Before:
Screenshot 2023-06-19 at 15 33 48

After:
Screenshot 2023-06-19 at 15 29 22

### CODE REVIEW

File: apps/web/config-overrides.js [hunk 1]
```diff
@@ -1,3 +1,10 @@
 const { useBabelRc, override } = require('customize-cra');
+// const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
```
Comment (p-fernandez) [inline apps/web/config-overrides.js:2]: I'd remove the comments before shipping it.

File: packages/notification-center/rollup.config.mjs [hunk 2]
```diff
@@ -8,33 +8,53 @@ import replace from '@rollup/plugin-replace';
 import gzipPlugin from 'rollup-plugin-gzip';
 import nodePolyfills from 'rollup-plugin-polyfill-node';
 import peerDepsExternal from 'rollup-plugin-peer-deps-external';
-import packageJson from './package.json' assert { type: 'json' };
+// import bundleAnalyzer from 'rollup-plugin-bundle-analyzer';
+
+function onwarn(warning, warn) {
```
Comment (p-fernandez) [inline packages/notification-center/rollup.config.mjs:13]:
```suggestion
function onWarn(warning, warn) {
```

**Review by p-fernandez — APPROVED** [review]
🌟

File: packages/notification-center/rollup.config.mjs [hunk 3]
```diff
@@ -8,33 +8,53 @@ import replace from '@rollup/plugin-replace';
...
     input: 'src/index.ts',
     output: [
       {
-        file: packageJson.main,
+        dir: 'dist/cjs',
         format: 'cjs',
-        sourcemap: true,
         interop: 'auto',
+        preserveModules: true, // indicate not create a single-file
+        preserveModulesRoot: 'src', // optional but useful to create a more plain folder structure
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:32]: preserve original folder structure

File: packages/notification-center/rollup.config.mjs [hunk 4]
```diff
@@ -8,33 +8,53 @@ import replace from '@rollup/plugin-replace';
...
       {
-        file: packageJson.module,
+        dir: 'dist/esm',
         format: 'esm',
+        preserveModules: true,
+        preserveModulesRoot: 'src',
         sourcemap: true,
       },
     ],
+    treeshake: 'smallest',
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:43]: apply the tree shaking

File: packages/notification-center/rollup.config.mjs [hunk 5]
```diff
@@ -8,33 +8,53 @@ import replace from '@rollup/plugin-replace';
...
       nodeExternals(),
       resolve({ preferBuiltins: false, browser: true }),
-      commonjs(),
       typescript({ tsconfig: './tsconfig.json' }),
-      terser(),
+      commonjs(),
+      replace({
+        preventAssignment: true,
+        'process.env.NODE_ENV': JSON.stringify('production'),
+      }),
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:53]: the order of all plugins matters

File: packages/notification-center/rollup.config.mjs [hunk 6]
```diff
@@ -8,33 +8,53 @@ import replace from '@rollup/plugin-replace';
...
       image(),
+      // bundleAnalyzer({ analyzerMode: 'static' }),
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:55]: left for the future to debug the NC bundle size

File: packages/notification-center/rollup.config.mjs [hunk 7]
```diff
@@ -46,18 +66,21 @@ export default [
         sourcemap: true,
       },
     ],
+    treeshake: 'smallest',
     plugins: [
+      resolve({ browser: true }),
+      nodePolyfills(),
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:72]: the node polyfills are needed for the Axios

File: packages/notification-center/rollup.config.mjs [hunk 8]
```diff
@@ -46,18 +66,21 @@ export default [
         sourcemap: true,
       },
     ],
+    treeshake: 'smallest',
     plugins: [
+      resolve({ browser: true }),
```
Comment (LetItRock) [inline packages/notification-center/rollup.config.mjs:71]: for Rollup to locate and bundle third-party dependencies in node_modules

File: packages/notification-center/src/components/notification-center/components/Loader.tsx [hunk 9]
```diff
@@ -1,9 +1,7 @@
 import React from 'react';
 import { Loader as MantineLoader } from '@mantine/core';
 import { cx, css } from '@emotion/css';
-import chroma from 'chroma-js';
```
Comment (LetItRock) [inline packages/notification-center/src/components/notification-center/components/Loader.tsx:4]: removed `chroma-js` it was used to translate the `linear-gradient()` to an average hex color, so instead I've calculated that color for the NC and used it in the theme config... the weight of that lib was ~18kB gzip, ~45kB unpacked... no sense to use library for that

File: packages/notification-center/src/components/notification-center/components/NotificationsListTab.tsx [hunk 10]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import { IMessage, ChannelCTATypeEnum } from '@novu/shared';
 
 import { useNotifications, useNotificationCenter, useNovuContext, useTranslations } from '../../../hooks';
-import image from '../../../images/no-new-notifications.png';
+import image from '../../../images/no-new-notifications.webp';
```
Comment (LetItRock) [inline packages/notification-center/src/components/notification-center/components/NotificationsListTab.tsx:5]: changed images to webp which is supported by all modern browsers: https://caniuse.com/?search=webp

File: packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx [hunk 11]
```diff
@@ -3,7 +3,6 @@ import { Avatar as MAvatar, ActionIcon, Menu, createStyles, MantineTheme } from
 import { useHover } from '@mantine/hooks';
 import { css, cx } from '@emotion/css';
 import styled from '@emotion/styled';
-import { formatDistanceToNow } from 'date-fns';
```
Comment (LetItRock) [inline packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx:6]: removed the `date-fns` library and all it's locales... instead we do use now built-in `Intl.RelativeTimeFormat`

File: packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx [hunk 12]
```diff
@@ -186,7 +186,7 @@ export function NotificationListItem({
                 css(listItemTimestampStyles)
               )}
             >
-              {formatDistanceToNow(new Date(notification.createdAt), { addSuffix: true, locale: dateFnsLocale() })}
+              {formatRelativeTime({ fromDate: new Date(notification.createdAt), locale: lang })}
```
Comment (LetItRock) [inline packages/notification-center/src/components/notification-center/components/notification-item/NotificationListItem.tsx:189]: our function that does the same, calculates the time passed from the date to the current date

File: packages/notification-center/src/shared/config/colors.ts [hunk 13]
```diff
@@ -16,6 +16,7 @@ export const colors = {
   error: '#E54545',
   vertical: 'linear-gradient(0deg,#FF512F 0%,#DD2476 100%)',
   horizontal: 'linear-gradient(99deg,#DD2476 0% 0%,#FF512F 100% 100%)',
+  loader: '#ef3f5a',
```
Comment (LetItRock) [inline packages/notification-center/src/shared/config/colors.ts:19]: the calculated loader color based on the gradient `vertical`

File: packages/notification-center/src/utils/date.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+  month: YEAR / 12,
+  day: DAY,
+  hour: HOUR,
+  minute: MINUTE,
+  second: SECOND,
+};
+
+const DEFAULT_LOCALE = 'en';
+
+export function formatRelativeTime({
```
Comment (LetItRock) [inline packages/notification-center/src/utils/date.ts:18]: the similar function that `date-fns` provides `formatDistanceToNow`, but this one uses the `Intl.RelativeTimeFormat`

File: packages/notification-center/tsconfig.json [hunk 15]
```diff
@@ -3,17 +3,17 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "forceConsistentCasingInFileNames": true,
-    "target": "es5",
+    "target": "es6",
```
Comment (LetItRock) [inline packages/notification-center/tsconfig.json:6]: All modern browsers fully support ES6: https://kangax.github.io/compat-table/es6/
TS docs: https://www.typescriptlang.org/tsconfig#target

File: packages/notification-center/rollup.config.mjs (hunk 2 above)
Comment (p-fernandez) [reply packages/notification-center/rollup.config.mjs:13]: No need, thought was a typo.

**Pull request merged**